xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLbIsReady {
+			gsb.swap()
+		}
+	} else {
+		gsb.currentLbIsReady = state.ConnectivityState == connectivity.Ready
+		// In the specific case that the current lb exits ready, and there is a
+		// pending lb, you can forward the pending LB's cached State up to
+		// ClientConn and swap the pending into the current. "Otherwise, the
+		// channel will keep using the old policy until...the old policy exits
+		// READY." - Java
+		if !gsb.currentLbIsReady && gsb.balancerPending != nil {
+			gsb.swap()
+		} else {
+			// Even if there is a pending balancer waiting to be gracefully
+			// switched to, still forward current balancer updates to the Client
+			// Conn. Ignoring state + picker from the current would cause
+			// undefined behavior/cause the system to behave incorrectly from
+			// the current LB policies perspective. Also, the current LB is
+			// still being used by grpc to choose SubConns per RPC, and thus
+			// should use the most updated form of the current balancer.
+			gsb.cc.UpdateState(state)
+		}
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	finished := make(chan struct{})
+	go func() {
+		gsb.swapMu.Lock()
+		defer gsb.swapMu.Unlock()
+		gsb.balRecentlyClosed = gsb.balancerCurrent
+		gsb.balancerCurrent.Close()
+		for sc, bal := range gsb.scToSubBalancer {
+			if bal == gsb.balancerCurrent {
+				gsb.cc.RemoveSubConn(sc) // Or does this happen implicitly?
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+
+		gsb.balancerCurrent = gsb.balancerPending
+		gsb.balancerPending = nil
+		finished <- struct{}{}
+	}()
+	<-finished
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bal balancer.Balancer) bool {
+	return bal == gsb.balancerCurrent || bal == gsb.balancerPending
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if gsb.closed.HasFired() {
+		// logger?
+		return nil, errBalancerClosed
+	}
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bal, bal)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+// Eat newSubConn calls from current if there is a pending?
+
+// Eat NewAddresses calls from current if there is a pending? (Allowing ClientConn to do work that doesn't need)
```

Comment:
Eat NewAddresses calls **TO** current

And is this comment in the right place?

Reply:
Deleted this comment.